balloon/src/lib.rs

```diff
@@ -0,0 +1,346 @@
+    pub fn hash_with_memory(
+        &self,
+        pwd: &[u8],
+        salt: &[u8],
+        memory_blocks: &mut [GenericArray<u8, D::OutputSize>],
+    ) -> Result<GenericArray<u8, D::OutputSize>> {
+        if self.params.p_cost.get() > 1 {
+            return Err(Error::ThreadsTooMany);
+        }
+
+        let s_cost = usize::try_from(self.params.s_cost.get()).unwrap();
+        let t_cost = usize::try_from(self.params.t_cost.get()).unwrap();
+
+        let mut digest = D::new();
+
+        // This is a direct translation of the `Balloon` from <https://eprint.iacr.org/2016/027.pdf> chapter 3.1.
+        // int delta = 3 // Number of dependencies per block
+        const DELTA: u32 = 3;
+        // int cnt = 0 // A counter (used in security proof)
+        let mut cnt: u32 = 0;
+        // block_t buf[s_cost]): // The main buffer
+        let memory_blocks = memory_blocks
+            .get_mut(..s_cost)
+            .ok_or(Error::MemoryTooLittle)?;
+
+        // Step 1. Expand input into buffer.
+        // buf[0] = hash(cnt++, passwd, salt)
+        digest.update(cnt.to_le_bytes());
+        cnt += 1;
+        digest.update(pwd);
+        digest.update(salt);
+
+        if let Some(secret) = self.secret {
+            digest.update(secret);
+        }
+
+        memory_blocks[0] = digest.finalize_reset();
+
+        // for m from 1 to s_cost-1:
+        for m in 1..s_cost {
+            // buf[m] = hash(cnt++, buf[m-1])
+            digest.update(&cnt.to_le_bytes());
+            cnt += 1;
+            digest.update(&memory_blocks[m - 1]);
+            memory_blocks[m] = digest.finalize_reset();
+        }
+
+        // Step 2. Mix buffer contents.
+        // for t from 0 to t_cost-1:
+        for t in 0..t_cost {
+            // for m from 0 to s_cost-1:
+            for m in 0..s_cost {
+                // Step 2a. Hash last and current blocks.
+                // block_t prev = buf[(m-1) mod s_cost]
+                let prev = if m == 0 { s_cost - 1 } else { m - 1 };
+                let prev = &memory_blocks[prev];
+
+                // buf[m] = hash(cnt++, prev, buf[m])
+                digest.update(&cnt.to_le_bytes());
+                cnt += 1;
+                digest.update(prev);
+                digest.update(&memory_blocks[m]);
+                memory_blocks[m] = digest.finalize_reset();
+
+                // Step 2b. Hash in pseudorandomly chosen blocks.
+                // for i from 0 to delta-1:
+                for i in 0..DELTA {
+                    // block_t idx_block = ints_to_block(t, m, i)
+                    // int other = to_int(hash(cnt++, salt, idx_block)) mod s_cost
+                    digest.update(&cnt.to_le_bytes());
+                    cnt += 1;
+                    digest.update(salt);
+                    digest.update(&u32::try_from(t).unwrap().to_le_bytes());
+                    digest.update(&u32::try_from(m).unwrap().to_le_bytes());
+                    digest.update(&i.to_le_bytes());
```

Comment:
I don't know what `ints_to_block` is supposed to mean, the reference implementation does something with AES (I think) that I didn't understand: [libballoon/hash_state.c#L171](https://github.com/henrycg/balloon/blob/0ae35a8879a97f4c7c44ee7fe77ea6522c607ef7/libballoon/hash_state.c#L171) -> [libballoon/bitstream.c#L152](https://github.com/henrycg/balloon/blob/0ae35a8879a97f4c7c44ee7fe77ea6522c607ef7/libballoon/bitstream.c#L152) -> [libballoon/bitstream.c#L134](https://github.com/henrycg/balloon/blob/0ae35a8879a97f4c7c44ee7fe77ea6522c607ef7/libballoon/bitstream.c#L134) -> [libballoon/bitstream.c#L118](https://github.com/henrycg/balloon/blob/0ae35a8879a97f4c7c44ee7fe77ea6522c607ef7/libballoon/bitstream.c#L118) -> [EVP_EncryptUpdate](https://www.openssl.org/docs/man1.1.1/man3/EVP_EncryptUpdate.html)

I took some inspiration from [a Python implementation](https://github.com/nachonavarro/balloon-hashing/blob/5e1eb7acf008428b7cca96fa5dbb9a481bed2eef/balloon.py#L78) instead.

Reply:
Also not an expert, but ints_to_block seems to just be for type safety than an actual cryptographic operation. Feeding directly into the hash seems fine

Reply:
Alright then.